src/TurbulenceClosures/turbulence_closure_implementations/Smagorinskys/dynamic_coefficient.jl

```diff
@@ -0,0 +1,248 @@
+        Σ[i, j, k] = √(ΣᵢⱼΣᵢⱼᶜᶜᶜ(i, j, k, grid, u, v, w))
+        Σ̄[i, j, k] = √(Σ̄ᵢⱼΣ̄ᵢⱼᶜᶜᶜ(i, j, k, grid, u, v, w))
+    end
+end
+
+@kernel function _compute_LM_MM!(LM, MM, Σ, Σ̄, grid, u, v, w)
+    i, j, k = @index(Global, NTuple)
+    LM_ijk, MM_ijk = LM_and_MM(i, j, k, grid, Σ, Σ̄, u, v, w)
+    @inbounds begin
+        LM[i, j, k] = LM_ijk
+        MM[i, j, k] = MM_ijk
+    end
+end
+
+@inline function LM_and_MM(i, j, k, grid, Σ, Σ̄, u, v, w)
+    L₁₁ = L₁₁ᶜᶜᶜ(i, j, k, grid, u, v, w)
+    L₂₂ = L₂₂ᶜᶜᶜ(i, j, k, grid, u, v, w)
+    L₃₃ = L₃₃ᶜᶜᶜ(i, j, k, grid, u, v, w)
+    L₁₂ = L₁₂ᶜᶜᶜ(i, j, k, grid, u, v, w)
+    L₁₃ = L₁₃ᶜᶜᶜ(i, j, k, grid, u, v, w)
+    L₂₃ = L₂₃ᶜᶜᶜ(i, j, k, grid, u, v, w)
+
+    M₁₁ = M₁₁ᶜᶜᶜ(i, j, k, grid, u, v, w, 2, 1, Σ, Σ̄)
+    M₂₂ = M₂₂ᶜᶜᶜ(i, j, k, grid, u, v, w, 2, 1, Σ, Σ̄)
+    M₃₃ = M₃₃ᶜᶜᶜ(i, j, k, grid, u, v, w, 2, 1, Σ, Σ̄)
+    M₁₂ = M₁₂ᶜᶜᶜ(i, j, k, grid, u, v, w, 2, 1, Σ, Σ̄)
+    M₁₃ = M₁₃ᶜᶜᶜ(i, j, k, grid, u, v, w, 2, 1, Σ, Σ̄)
+    M₂₃ = M₂₃ᶜᶜᶜ(i, j, k, grid, u, v, w, 2, 1, Σ, Σ̄)
+
+    LM_ijk = L₁₁ * M₁₁ + L₂₂ * M₂₂ + L₃₃ * M₃₃ + 2L₁₂ * M₁₂ + 2L₁₃ * M₁₃ + 2L₂₃ * M₂₃
+    MM_ijk = M₁₁ * M₁₁ + M₂₂ * M₂₂ + M₃₃ * M₃₃ + 2M₁₂ * M₁₂ + 2M₁₃ * M₁₃ + 2M₂₃ * M₂₃
+
+    return LM_ijk, MM_ijk
+end
+
+#####
+##### Directionally-averaged functionality
+#####
+
+function compute_coefficient_fields!(diffusivity_fields, closure::DirectionallyAveragedDynamicSmagorinsky, model; parameters)
+    grid = model.grid
+    arch = architecture(grid)
+    velocities = model.velocities
+    cˢ = closure.coefficient
+
+    if cˢ.schedule(model)
+        Σ = diffusivity_fields.Σ
+        Σ̄ = diffusivity_fields.Σ̄
+        launch!(arch, grid, :xyz, _compute_Σ_Σ̄!, Σ, Σ̄, grid, velocities...)
+
+        LM = diffusivity_fields.LM
+        MM = diffusivity_fields.MM
+        launch!(arch, grid, :xyz, _compute_LM_MM!, LM, MM, Σ, Σ̄, grid, velocities...)
+
+        𝒥ᴸᴹ = diffusivity_fields.𝒥ᴸᴹ
+        𝒥ᴹᴹ = diffusivity_fields.𝒥ᴹᴹ
+        compute!(𝒥ᴸᴹ)
+        compute!(𝒥ᴹᴹ)
+    end
+
+    return nothing
+end
+
+function allocate_coefficient_fields(closure::DirectionallyAveragedDynamicSmagorinsky, grid)
+    LM = CenterField(grid)
+    MM = CenterField(grid)
+
+    Σ = CenterField(grid)
+    Σ̄ = CenterField(grid)
+
+    𝒥ᴸᴹ = Field(Average(LM, dims=closure.coefficient.averaging))
+    𝒥ᴹᴹ = Field(Average(MM, dims=closure.coefficient.averaging))
+
+    return (; Σ, Σ̄, LM, MM, 𝒥ᴸᴹ, 𝒥ᴹᴹ)
+end
+
+#####
+##### Lagrangian-averaged functionality
+#####
+
+const c = Center()
+
+@kernel function _lagrangian_average_LM_MM!(𝒥ᴸᴹ, 𝒥ᴹᴹ, 𝒥ᴸᴹ⁻, 𝒥ᴹᴹ⁻, 𝒥ᴸᴹ_min, Σ, Σ̄, grid, Δt, u, v, w)
+    i, j, k = @index(Global, NTuple)
+    LMⁿ, MMⁿ = LM_and_MM(i, j, k, grid, Σ, Σ̄, u, v, w)
+    FT = eltype(grid)
+
+    @inbounds begin
+        𝒥ᴸᴹ⁻ᵢ = max(𝒥ᴸᴹ⁻[i, j, k], 𝒥ᴸᴹ_min)
+        𝒥ᴹᴹ⁻ᵢ = 𝒥ᴹᴹ⁻[i, j, k]
+
+        # Compute time scale
+        𝒥ᴸᴹ𝒥ᴹᴹ = 𝒥ᴸᴹ⁻ᵢ * 𝒥ᴹᴹ⁻ᵢ
+
+        if 𝒥ᴸᴹ𝒥ᴹᴹ < 0
+            @show 𝒥ᴸᴹ⁻ᵢ 𝒥ᴹᴹ⁻ᵢ 
+        end
```

Comment:
@glwagner I'm assuming I can remove this?

Reply:
yes!